src/adapter/src/client.rs

```diff
@@ -385,157 +416,251 @@ impl SessionClient {
+                    Some(desc) => desc.iter_names().map(|name| name.to_string()).collect(),
+                    None => vec![],
+                };
+                let mut datum_vec = mz_repr::DatumVec::new();
+                for row in rows {
+                    let datums = datum_vec.borrow_with(&row);
+                    sql_rows.push(datums.iter().map(From::from).collect());
                 }
-                | ExecuteResponse::Updated(_)
-                | ExecuteResponse::AlteredObject(_)
-                | ExecuteResponse::AlteredIndexLogicalCompaction
-                | ExecuteResponse::AlteredSystemConfiguraion
-                | ExecuteResponse::Deallocate { all: _ }
-                | ExecuteResponse::Prepare) => {
-                    results.push(SimpleResult::ok(res));
+                Ok(SimpleResult::Rows {
+                    rows: sql_rows,
+                    col_names,
+                })
+            }
+            ExecuteResponse::Fetch { .. }
+            | ExecuteResponse::SetVariable { .. }
+            | ExecuteResponse::Tailing { .. }
+            | ExecuteResponse::CopyTo { .. }
+            | ExecuteResponse::CopyFrom { .. }
+            | ExecuteResponse::Raise { .. }
+            | ExecuteResponse::DeclaredCursor
+            | ExecuteResponse::ClosedCursor => {
+                // NOTE(benesch): it is a bit scary to ignore the response
+                // to these types of planning *after* they have been
+                // planned, as they may have mutated state. On a quick
+                // glance, though, ignoring the execute response in all of
+                // the above situations seems safe enough, and it's a more
+                // target allowlist than the code that was here before.
+                Ok(SimpleResult::err(
+                    "executing statements of this type is unsupported via this API",
+                ))
+            }
+        }
+    }
+
+    /// Executes an [`HttpSqlRequest`].
+    ///
+    /// The provided `stmts` are executed directly, and their results
+    /// are returned as a vector of rows, where each row is a vector of JSON
+    /// objects.
+    pub async fn execute_sql_http_request(
+        &mut self,
+        request: HttpSqlRequest,
+    ) -> Result<SqlHttpExecuteResponse, AdapterError> {
+        let mut results = vec![];
+
+        match request {
+            HttpSqlRequest::Simple { query } => {
+                let stmts = mz_sql::parse::parse(&query)
+                    .map_err(|e| AdapterError::Unstructured(e.into()))?;
+                let num_stmts = stmts.len();
+
+                for stmt in stmts {
+                    // Mirror the behavior of the PostgreSQL simple query protocol.
+                    // See the pgwire::protocol::StateMachine::query method for details.
+                    self.start_transaction(Some(num_stmts)).await?;
+                    results.push(self.execute_stmt(stmt, vec![]).await?);
```

Comment:
Do both the simple and extended variants need to stop execution on first error?

Reply:
`execute_stmt` returns an `Err` sometimes, which would cause this function to return and show only that error to the user. But a query like `insert into t values (1); commit; select $1` would commit the insert and then return an error to the user, without them being told the insert committed.

Reply:
re:

> Do both the simple and extended variants need to stop execution on first error?

I believe the extended variant does stop processing, as well, per [this line](https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY):

> At completion of each series of extended-query messages, the frontend should issue a Sync message. This parameterless message causes the backend to close the current transaction if it's not inside a BEGIN/COMMIT transaction block (“close” meaning to commit if no error, or roll back if error). Then a ReadyForQuery response is issued. The purpose of Sync is to provide a resynchronization point for error recovery. When an error is detected while processing any extended-query message, the backend issues ErrorResponse, then reads and discards messages until a Sync is reached, then issues ReadyForQuery and returns to normal message processing.

So it sounds like if it issues an error, it discards everything until a `Sync`, which in this case I believe we're simulating when the queries in the request have all been processed.

Re: the second comment, I mentioned this in [this comment](https://github.com/MaterializeInc/materialize/pull/14683#issuecomment-1239510058), but it's moot now because I merged https://github.com/MaterializeInc/materialize/pull/14582 and am rebasing it on top of this, which will include a fix for the bad error handling.